src/main/java/org/folio/rest/impl/EholdingsProvidersImpl.java

```diff
@@ -0,0 +1,96 @@
+import org.folio.config.api.RMAPIConfigurationService;
+import org.folio.http.ConfigurationClientProvider;
+import org.folio.rest.annotations.Validate;
+import org.folio.rest.converter.VendorConverter;
+import org.folio.rest.jaxrs.model.EholdingsProvidersPackagesGetFilterSelected;
+import org.folio.rest.jaxrs.model.EholdingsProvidersPackagesGetFilterType;
+import org.folio.rest.jaxrs.model.EholdingsProvidersProviderIdGetInclude;
+import org.folio.rest.jaxrs.model.ProviderPutRequest;
+import org.folio.rest.jaxrs.resource.EholdingsProviders;
+import org.folio.rest.model.OkapiData;
+import org.folio.rest.util.ErrorUtil;
+import org.folio.rest.validator.HeaderValidator;
+import org.folio.rmapi.RMAPIService;
+
+import javax.ws.rs.core.Response;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+
+public class EholdingsProvidersImpl implements EholdingsProviders {
+
+  private static final String GET_PROVIDERS_ERROR_MESSAGE = "Failed to retrieve providers";
+
+  private final Logger logger = LoggerFactory.getLogger(EholdingsConfigurationImpl.class);
+
+  private RMAPIConfigurationService configurationService;
+  private HeaderValidator headerValidator;
+  private VendorConverter converter;
+
+  public EholdingsProvidersImpl() {
+    this(
+      new RMAPIConfigurationServiceCache(
+        new RMAPIConfigurationServiceImpl(new ConfigurationClientProvider())),
+      new HeaderValidator(),
+      new VendorConverter());
+  }
+
+  public EholdingsProvidersImpl(RMAPIConfigurationService configurationService,
+                                HeaderValidator headerValidator,
+                                VendorConverter converter) {
+    this.configurationService = configurationService;
+    this.headerValidator = headerValidator;
+    this.converter = converter;
+  }
+
+  @Override
+  @Validate
+  public void getEholdingsProviders(String q, int page, int count, String sort, Map<String, String> okapiHeaders, Handler<AsyncResult<Response>> asyncResultHandler, Context vertxContext) {
+    if (!headerValidator.validate(okapiHeaders, asyncResultHandler)) {
+      return;
+    }
+    CompletableFuture.completedFuture(null)
+      .thenCompose(o -> configurationService.retrieveConfiguration(new OkapiData(okapiHeaders)))
+      .thenCompose(rmapiConfiguration -> {
+        RMAPIService rmapiService = new RMAPIService(rmapiConfiguration.getCustomerId(), rmapiConfiguration.getAPIKey(),
+          rmapiConfiguration.getUrl(), vertxContext.owner());
+        return rmapiService.retrieveProviders(q, page, count, sort);
+      })
+      .thenAccept(vendors ->
+        asyncResultHandler.handle(Future.succeededFuture(GetEholdingsProvidersResponse
+        .respond200WithApplicationVndApiJson(converter.convert(vendors)))))
+      .exceptionally(e -> {
+        logger.error(GET_PROVIDERS_ERROR_MESSAGE, e);
+        asyncResultHandler.handle(Future.succeededFuture(GetEholdingsProvidersResponse
+          .status(500)
```

Comment:
There are certain validations we do in the Ruby code and give 400 with an appropriate error message. Also, sometimes, when we get an error message from RM API, we pass that error message through so its useful to the end-user. Example: If the user makes a request with count=120, then we give a 400 with errors.title being `Parameter Count is outside the range 1-100`, where as here, it looks like we are giving a 500 for any exception.

Reply:
There is validation of parameters in mod-kb-ebsco-java module, so if parameter is invalid mod-kb-ebsco-java will imediately return a 400 error, without sending request to RM API. 
Validation is implemented by using @Validate annotation, parameters will be automatically validated against interface generated from raml (e.g. count parameter is generated like this "@QueryParam("count") @DefaultValue("25") @Min(0) @Max(100) int count", so if count is greater than 100 an error will be returned),
here's an example error
```
{
    "errors": [
        {
            "detail": "\n 'count' parameter is incorrect. parameter value {1001} is not valid: must be less than or equal to 100"
        }
    ],
    "jsonapi": {
        "version": "1.0"
    }
}
```

But it is still nice to return RM API error message in case of other errors.
I changed RMAPIService to include response body from RM API. That way it will be included in 500 response:
```
{
    "errors": [
        {
            "detail": "Unauthorized Access to https://sandbox.ebsco.io/rm/rmaccounts/apidvgvmt/vendors?search=Cengage&offset=1&count=10&orderby=relevance, response body =\"{\"message\":\"Unauthorized\"}\n\""
        }
    ],
    "jsonapi": {
        "version": "1.0"
    }
}
```

Reply:
Is there a way we can customize the error messages we want to return using @validate annotation? The reason I ask is, in mod-kb-ebsco, we return certain customized error messages and populate them in errors.title which is what is displayed in UI as a toast message.
Also, regarding 500's, I think it will be good to differentiate status codes based on type of exceptions and return them appropriately, like 403 probably for a "unauthorized access" and again, not sure if that is the exact error message being returned from mod-kb. Thanks!

Reply:
let's address this in another US